### ISSUE TITLE
Callback Link Not Working (Projects + Reviews)

### ISSUE THREAD

**Issue by CraneRewind** [NotePlan/plugins#509]
Same behaviour as https://github.com/NotePlan/plugins/issues/508 but in a different plugin.

**Describe the bug**
In both rich and markdown view, clicking on a link results in no actions at all.

**To Reproduce**
1. Set up at least one project for the Projects + Reviews plugin to work with.
2. Trigger `/project lists`
3. I have both rich+markdown.  You should see a markdown note with links, a Review List window with links, or both.  (see setup screenshot)
4. Click links, either in markdown note or the rich dialog.  

**Expected behavior**
I expect something to happen when I click a link.  I expect to see some debug console action.

**Screenshots**
See attachments.

**System**
 - Device: Mac
 - OS: macOS14.0 (23A344)
 - NotePlan Version: 3.9.11
 - Plugin Name & Version: Projects + Reviews v0.13.0

**Additional context**
I don't often use callbacks like this.  One plug was working perfect fine sometime last year, but I have no idea when the functionality broke.

**Plugin Console Log**
Not including the log because nothing happens.  You can see that in the video.

**Attachments**
![1 setup](https://github.com/NotePlan/plugins/assets/98125767/5150c57a-7f44-49f1-af8c-28aba049026e)
![2 debug set](https://github.com/NotePlan/plugins/assets/98125767/688e6516-009e-4dea-9bdf-4236bc0d6ff7)

https://github.com/NotePlan/plugins/assets/98125767/28f217e1-0579-4906-ae76-7cdcc3e418bc

**Comment by jgclark** [comment]
I cannot reproduce this, with the link you clicked on.
Are all the other buttons (rich style), or links (markdown style) also not doing anything?

From the screenshot you showed on Discord, the x-callback looked like it might have a space in it. It should read:
`noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=start%20reviews`
That's what I have, but if there's a space instead of the `%20` nothing will happen. Please could you ensure that on the Markdown version it has the `%20`, and then try it again?

**Comment by EduardMe** [comment]
Have just tried the link on my side (without any other setup) and it worked for me. Just showed a message that there's nothing to review. Does this work for you? Means you delete the reviewable notes and then try it?

**Comment by jgclark** [comment]
The problem is stopping the plugin command even starting. So there's something funny in your app or the plugin installation. I suggest in order:
- restarting app
- deleting and reinstalling plugin (settings will be saved)
- delete and reinstall app

**Comment by CraneRewind** [comment]
JG, I've tried, but no joy.

Note: other plugins seem to work fine, e.g.,

- AutoTimeBlockingEvents (insert today as list)
- Dashboard (refresh and toggle buttons)

^^ not sure if either of those is using callbacks

**Comment by EduardMe** [comment]
The beta that interprets empty values in x-callback-urls properly as undefined is up and we can test this. But not sure if it helps here.

Did we already enabled logging by the way? https://help.noteplan.co/article/57-how-to-enable-logging

**Comment by CraneRewind** [comment]
Oh, that logging is interesting.  This is what I did:

- deleted the old log files
- turned on logging in noteplan
- restarted noteplan
- typed `/projects` and then selected /projects lists
- clicked the refresh button a couple of times
- the dialog didn't show anything, so ran /projects lists again
- now the dialog shows something
- clicked review
- clicked cancel
[np-error.log](https://github.com/NotePlan/plugins/files/13878213/np-error.log)
[np-out.log](https://github.com/NotePlan/plugins/files/13878214/np-out.log)

I see a lot of repeat messages (even when I started typing /pro) about "np.MeetingNotes is already downloaded"; that seems like a warning.  Also some bits about Dropbox, FYI, not using dropbox.

Not sure it should matter, but FYI, I don't sync my notes, dropbox and cloud are not allowed.

**Comment by EduardMe** [comment]
Indeed interesting, I'm adding more logging to show what's available to the next beta. I checked on my side and it can find everything. When I hit start review, it opens the Secret Undertaking note for me.

However, there's more plugin specific text in the log, maybe @jgclark has a clue what it means.

**Comment by EduardMe** [comment]
Small update with the latest beta we have more logging now to figure out why the plugin can't be found!

**Comment by jgclark** [comment]
@CraneRewind I can't tell from the messages whether this is now working fully or not for you.  But one thing I want to check from the logs is what your file extension for notes is. I'm guessing "txt" not "md"?

**Comment by jgclark** [comment]
More questions for you:
- Are you running the Setapp version?
- Does the file "/Users/crane/Library/Containers/co.noteplan.NotePlan-setapp/Data/Library/Application Support/co.noteplan.NotePlan-setapp/Plugins/data/jgclark.Dashboard/full-review-list.md" exist?  If so does it contain anything?
- Does the note '#project Review List' exist and show something useful?

**Comment by CraneRewind** [comment]
> Indeed interesting, I'm adding more logging to show what's available when the plugin detection fails to the next beta. I checked on my side and it can find everything. When I hit start review, it opens the Secret Undertaking note for me.
> 
> However, there's more plugin specific text in the log, maybe @jgclark has a clue what it means.
> 
> Also, if you haven't tried yet, reinstall the plugin (you probably did that already) and I would also try deleting all other plugins, but this one.

I did try uninstall/reinstall.  I've also tried deleting all the plugins (although I couldn't get rid of a couple.)

**Comment by CraneRewind** [comment]
> @CraneRewind I can't tell from the messages whether this is now working fully or not for you. But one thing I want to check from the logs is what your file extension for notes is. I'm guessing "txt" not "md"?

Not working yet.  My file extension is "txt".

**Comment by CraneRewind** [comment]
> More questions for you:
> 
> * Are you running the Setapp version?
> * Does the file "/Users/crane/Library/Containers/co.noteplan.NotePlan-setapp/Data/Library/Application Support/co.noteplan.NotePlan-setapp/Plugins/data/jgclark.Dashboard/full-review-list.md" exist?  If so does it contain anything?
> * Does the note '#project Review List' exist and show something useful?

- Yes, from Setapp.
- "/Users/crane/Library/Containers/co.noteplan.NotePlan-setapp/Data/Library/Application Support/co.noteplan.NotePlan-setapp/Plugins/data/jgclark.Dashboard/full-review-list.md" exists and contains settings.json.  Content below.
- That note does exist.  Content below.


`{
  "includeTaskContext" : false,
  "showYesterdaySection" : true,
  "dashboardTheme" : "",
  "autoAddTrigger" : false,
  "ignoreTasksWithPhrase" : "",
  "excludeTasksWithTimeblocks" : false,
  "ignoreChecklistItems" : true,
  "triggerLogging" : false,
  "tagToShow" : "@waiting",
  "maxTasksToShowInSection" : 20,
  "ignoreFolders" : [
    "Saved Searches"
  ],
  "separateSectionForReferencedNotes" : true,
  "showWeekSection" : true,
  "showOverdueTaskSection" : true,
  "_logLevel" : "DEBUG",
  "overdueSortOrder" : "priority",
  "updateOverdueOnTrigger" : true,
  "includeFolderName" : true,
  "excludeChecklistsWithTimeblocks" : false
}`


`# #project Review List
Total 1 active projects: **[Start reviewing 1 ready for review](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=start%20reviews)** (showing all open projects/areasplus finished ones.) Last updated: 2024-01-09 2:38 PM [🔄 Refresh](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=project%20lists&arg0=noteTypeTags%253D%2523project)
Review: [Finish](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=finish%20project%20review) [Finish + Next](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=next%20project%20review) Current open project note: [Add progress](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=add%20progress%20update) [toggle Pause](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=pause%20project%20toggle) [Complete](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=complete%20project) [Cancel](noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=cancel%20project)
### 1 Projects
- [[Secret Undertaking]]	 50% done (of 4 tasks)	due in 23 days	Review due **today**`

**Comment by jgclark** [comment]
I've tried again to reproduce this and can't.
Is this still happening for you? 
If so please send contents of "/Users/crane/Library/Containers/co.noteplan.NotePlan-setapp/Data/Library/Application Support/co.noteplan.NotePlan-setapp/Plugins/data/jgclark.Dashboard/full-review-list.md" file.  (Note: last time you looked in the wrong directory. This has '.../data/...' in it.)

**Comment by CraneRewind** [comment]
Hi again JG.  Still not working, although I'm getting better at the / commands ;)

It looks like there's more info being logged now in the np-out.log when click one of the buttons.  Here, I've clicked the "Refresh" button in the Review List modal.

```
AppDelegate: handleGetURLEvent called, picking the first url: noteplan://x-callback-url/runPlugin?pluginID=jgclark.Reviews&command=project%20lists. Is the handler nil = no (cache the url if needed)
UI+URLScheme: parse incoming x-callback-url
UI+URLScheme: do something with runPlugin
Running a plugin through x-callback (["pluginID": "jgclark.Reviews", "command": "project lists"]
Plugin 'np.Templating' is already downloaded, don't add it from the core plugins again
Coudn't run plugin, need `pluginID` or `pluginName` as parameter, parameters: ["pluginID": "jgclark.Reviews", "command": "project lists"]
Plugin 'np.Templating' is already downloaded, don't add it from the core plugins again
Available plugins: ["Name: 🗓 AutoTimeBlocking / Events, ID: dwertheimer.EventAutomations", "Name: 🎛 Dashboard, ID: jgclark.Dashboard", "Name: 🕓 Event Helpers, ID: jgclark.EventHelpers", "Name: ⭐️ Favorites, ID: dwertheimer.Favorites", "Name: ✍️ Meeting Notes, ID: np.MeetingNotes", "Name: 🔢 Note Statistics, ID: np.statistics", "Name: ⬇️ Plugin Installer, ID: np.installer", "Name: ⚡️ Quick Capture, ID: jgclark.QuickCapture", "Name: 🤝 Shared Resources, ID: np.Shared", "Name: 📒 Templating, ID: np.Templating", "Name: 🧹 Tidy Up, ID: np.Tidy"]
```

I find it weird that
1. It wants pluginID and it looks like pluginID is in the params.
2. This plugins is not listed in the list of Available plugins.  Nor is the habit tracker.  I don't know if that's part of the issue, or they just don't show up because they don't show up in the Plugins menu drop-down.

I'm looking for that file.  No md in that particular directory.

```
~/L/Containers/co./D/L/Application Su/c/P/data/jgclark.Dashboard  pwd
/Users/crane/Library/Containers/co.noteplan.Noteplan-setapp/Data/Library/Application Support/co.noteplan.NotePlan-setapp/Plugins/data/jgclark.Dashboard

  07.03.24 14:11  0s  ✔ ----------------------------------------------------------------------------------------------------------- 3.0.6 
 ~/L/Containers/co./D/L/Application Su/c/P/data/jgclark.Dashboard  ls -la
   rwxr-xr-x    3   crane   staff     96 B     Mon May  1 16:13:05 2023    ./
   rwxr-xr-x   19   crane   staff    608 B     Fri Feb  2 16:02:21 2024    ../
   rw-r--r--    1   crane   staff    642 B     Wed Jan 10 16:03:54 2024    settings.json
```


but there is one in the .../jgClark.Reviews directory.  Contents

```
---
title: full-review-list
date: 2024-03-07T10:50:56-05:00
key: reviewDays	dueDays	title	folder	tags	state
---
2	NaN	P: xxx	1 Projects	#project 	active
31	NaN	P: xxx	1 Projects	#project 	active
7	NaN	P: xxx	1 Projects	#project 	active
0	NaN	A: xxx	2 Areas	#area 	active
24	NaN	A: xxx	2 Areas	#area 	active
0	NaN	A: xxx	2 Areas	#area 	active
0	NaN	A: xxx	2 Areas	#area 	active%
```

I had to elide the actual names though.

**Comment by jgclark** [comment]
@EduardMe the log from @CraneRewind in the previous post gives several leads:

- Confirms the callback is arriving in NP and being parsed
- Looks an error in logic (or perhaps unclear logging) given `Coudn't run plugin, need `pluginID` or `pluginName` as parameter, parameters: ["pluginID": "jgclark.Reviews", "command": "project lists"]`
- Unclear why several installed plugins aren't appearing in `Available plugins: [...]`

The fact that the plugin of interest here isn't in that list probably explains why the callbacks aren't doing anything. So the question is why does NP not have it in its list of installed ones?

**Comment by EduardMe** [comment]
@jgclark I have an idea. Might be these plugins were disabled? It's not showing disabled plugins in "Available plugins...", so NotePlan doesn't select it for the x-callback. I can set this to also scan disabled plugins and upload the beta again to test this.

Another way to test it, is to change the plugin ID, but that also messes with the x-callback-url, so better we wait for a beta upload. I just uploaded a beta without the changes, so will upload another tomorrow or so. 

@CraneRewind , is it possible you disabled the plugin at some point in time (we have removed the enable button some while ago)?

**Comment by CraneRewind** [comment]
@EduardMe yes, I likely disabled it, enabled it, uninstalled it, reinstalled it, etc. etc. trying to get it to work properly.

It shows as installed in the Preferences.
![image](https://github.com/NotePlan/plugins/assets/98125767/d38f216c-6256-4b69-b5ea-6b263d63af8a)

It is 'installed' in that it works, just not the callbacks.

Is there some file I can check/fix to have it show up in your list properly again?

**Comment by EduardMe** [comment]
It's difficult to access the preferences now. You could try to change the ID in the plugin.json file and adapt the callback as well, or wait and I'll upload a beta which ignores the disabled flag (we should have done that anyways). Here's a link to TestFlight, in case you are not on it. We're fixing some more thing today and upload this final beta ready for release (I hope): https://testflight.apple.com/join/fm9q4OjE

**Comment by CraneRewind** [comment]
Thanks Eduard!  I'm using NotePlan through Setapp, so I'll wait for the eventual fix.  It's nice to have a better idea of the "why" now.

**Comment by jgclark** [comment]
> @jgclark I have an idea. Might be these plugins were disabled? It's not showing disabled plugins in "Available plugins...", so NotePlan doesn't select it for the x-callback. I can set this to also scan disabled plugins and upload the beta again to test this.

Sorry, only just seen this. I've not done anything on my end to disable any plugins. And given its showing in his list of plugins, I'm not sure what you mean.  But you've since released a new beta: what did that change?

**Comment by EduardMe** [comment]
@CraneRewind I think this can also be closed now?

**Comment by CraneRewind** [comment]
Yes!  I've just installed the newest version (3.10.2) through Setapp and the links for my two plugins are working.  Excellent!  Thank you very much.
